File: solutions/Combination Sum/Combination Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        # Neetcode solution

        combinations = []

        def dfs(i, currCombo, total):
            # successful base case
            if total == target:
                combinations.append(currCombo[:]) # append a copy instead of a reference
                return

            # failed base case
            elif i >= len(candidates) or total > target:
                return

            # recursive call(s)
            else:
                currCombo.append(candidates[i])
                dfs(i, currCombo, total + candidates[i]) # branch to a duplicate of current candidate
                currCombo.pop()
                dfs(i+1, currCombo, total) # branch to the next candidate

        dfs(0,[],0)
        return combinations

File: solutions/Combination Sum/Combination Sum.ts
// NC solution from memory

function combinationSum(candidates: number[], target: number): number[][] {
    const allCombos: number[][] = [];
    const curCombo: number[] = [];
    dfs(0, 0);
    return allCombos;

    function dfs(i: number, curSum: number): void {
        // good base case
        if (curSum === target) {
            allCombos.push([...curCombo]);
            return;
        }
        // bad base case
        else if (i >= candidates.length || curSum > target) {
            return;
        }

        // O(2^n) recursion
        // include i in combo
        curCombo.push(candidates[i]);
        dfs(i, curSum+candidates[i]);

        // exclude i in combo
        curCombo.pop();
        dfs(i+1, curSum);
    }
};